### ISSUE TITLE
Frame rate is locked to 60FPS on some android devices 

### ISSUE THREAD

**Issue by Sreeraj9764** [flutter/flutter#160952]
### Steps to reproduce

1.Create new flutter app with the latest version
2.Check show refresh rate in developer options
3.Run your flutter app in devices that supports 120 or above
4.check refresh rate getting locked at 60

### Expected results

it should run at chosen refresh rate or at max, should not get locked at 60hz.

### Actual results

When any installed Flutter app is opened, it gets locked at 60Hz. In my case, the refresh rate drops from 120Hz to 60Hz. Only the Zerodha Kite app, which is made with Flutter, shows the proper refresh rate. It uses the [flutter_displaymode](https://pub.dev/packages/flutter_displaymode) package, which does not seem to work with the latest Flutter version.

### Code sample

<details open><summary>Code sample</summary>

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: Text('Hello, World!'),
        ),
      ),
    );
  }
}

</details>

### Screenshots or Video

This video recorded from Xiaomi Poco F5, also had same issues with Redmi,infinix etc

https://github.com/user-attachments/assets/8c116208-64cc-40af-9de4-283785541220

### Flutter Doctor output

<details open><summary>Doctor output</summary>

[✓] Flutter (Channel stable, 3.24.3, on macOS 14.6 23G80 darwin-arm64, locale en-IN)
    • Flutter version 3.24.3 on channel stable at /Users/growingstars/flutter/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (4 months ago), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/growingstars/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 21.0.3+-79915917-b509.11)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16A242d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 21.0.3+-79915917-b509.11)

[✓] VS Code (version 1.96.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.102.0

[✓] Connected device (3 available)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.6 23G80 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.6 23G80 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Google Chrome 131.0.6778.205

[✓] Network resources
    • All expected network resources are available.

• No issues found!

</details>

**Comment by Sreeraj9764** [comment]
this issue related to #35162 ,which seems to be closed after the plugin

**Comment by sheikhhaziq** [comment]
I was having the same issue and I resolved it by disabling impeller and using the plugin `flutter_displaymode`

**Comment by Sreeraj9764** [comment]
@sheikhhaziq 
I'm using Flutter version 3.24.3. Impeller became the default in version 3.27 for Android. Also, flutter_displaymode is not working for me. I think it depends on the Flutter SDK, not the device, because Kite is using version 3.19 and runs at a high refresh rate.

**Comment by darshankawar** [comment]
@Sreeraj9764 
It could be specific to the devices you mentioned and maybe related to specific hardware they are using in order to avoid over heating or even conserve battery usage. Check https://www.reddit.com/r/PocoPhones/comments/1c9067z/poco_f5_pro_fps_stuck_at_60/?rdt=34814 for further reference and see if it helps in your case or not.

**Comment by Sreeraj9764** [comment]
@darshankawar
Yes, it could be specific to devices. While debugging, it prints, _"dynamic FPS is applicable to this device"_ but I never get 120 FPS in debug mode. Additionally, there’s strange behaviour in release mode—it goes to 120 FPS only if I restart the device. My point is that there should be an option in the SDK itself to control this behaviour.

**Comment by darshankawar** [comment]
@Sreeraj9764 
Can you provide the device details in which you are seeing this behavior ? (make, model, OS name being used in them)

**Comment by rbozan** [comment]
I have the same issue, the app is stuck on 60Hz.

Model: Poco F5
Android 14
Xiaomi HyperOS 10.0.17.0UMREUXM

**Comment by rbozan** [comment]
> 小米设备：设置 > 应用设置 > 搜索“电量和性能” > 清除数据 > 重新打开应用。 这个问题应该是由小米系统的高刷策略导致的。

At least for me when I try this, initially it goes to 120Hz but after some time 60Hz will come back.

**Comment by SX-Code** [comment]
> At least for me when I try this, initially it goes to 120Hz but after some time 60Hz will come back.

在debug模式下无法维持120Hz，不过没关系。使用release模式，也就是`flutter build apk`生成的安装包可以维持120Hz。不过需要进行一次我提到的清除数据操作，它是根据包名决定是60Hz还是120Hz.不用太过烦恼。

**Comment by Sreeraj9764** [comment]
@SX-Code 

Yes, after using a Xiaomi device for debugging, we either have to clear the settings or restart the device to test in release mode. Otherwise, it stays at 60Hz instead of using 120Hz.

**Comment by darshankawar** [comment]
Keeping the issue open for team's attention as this seems device specific issue.

**Comment by n7trd** [comment]
I have the same issue with my phone (Nothing Phone 2a). It runs at 120Hz (not on the balanced setting!). 
All Flutter apps, except for the Wonderous App Demo, run at 60Hz, including Google Earth. 
The Wonderous App runs at 120Hz and feels very smooth. It's also the only app that uses the flutter_displaymode plugin according to FlutterShark app.

```dart
// Wonderous App Source
// Set preferred refresh rate to the max possible (the OS may ignore this)
if (!kIsWeb && PlatformInfo.isAndroid) {
      await FlutterDisplayMode.setHighRefreshRate();
}
```

It would be nice if that was the default for all Flutter apps because it causes a noticeable downgrade for all Flutter apps compared to native/React Native apps.

**Comment by Sreeraj9764** [comment]
@n7trd 
Is this issue occurring on your debugging device? Try installing the release build on another device that supports 120Hz and check if it works.

For your debugging device, try the following steps:

1. Uninstall the debug build.
2. Restart the device.
3. Install the release build.
4. Check if it runs at 120Hz.

**Comment by jonahwilliams** [comment]
On Android we the standard APIs for acquiring frames. If the platform doesn't give us frames at 120 fps, then we can't do it. Unclear why this is device specific. In the past we've seen issues were phones have denylists or use herustics like the presence of a SurfaceView to cut framerate - but that may not be the case here.

**Comment by n7trd** [comment]
@jonahwilliams 

I just tested the starter template app and can confirm that the app runs at only 60fps (device setting is on 120 fps).

After adding the flutter_displaymode plugin, it runs at 120fps (with FlutterDisplayMode.setHighRefreshRate()). 
So I cannot confirm, as pointed out above, that it's not working with Impeller.

Something I noticed in addition: I get these errors in the console:

```
D/BLASTBufferQueue(31877): [SurfaceView[com.example.testapp/com.example.testapp.MainActivity]#3](f:0,a:0) constructor()
D/BLASTBufferQueue(31877): [SurfaceView[com.example.testapp/com.example.testapp.MainActivity]#3](f:0,a:0) update width=1084 height=2412 format=4 mTransformHint=0
E/mali_gralloc(31877): ERROR: Format allocation info not found for format: 38
E/mali_gralloc(31877): ERROR: Format allocation info not found for format: 0
E/example.testapp(31877): Invalid base format! req_base_format = 0x0, req_format = 0x38
E/mali_gralloc(31877): ERROR: Unrecognized and/or unsupported format 0x38 and usage 0xb00
E/mali_gralloc(31877): ERROR: Format allocation info not found for format: 3b
E/mali_gralloc(31877): ERROR: Format allocation info not found for format: 0
E/example.testapp(31877): Invalid base format! req_base_format = 0x0, req_format = 0x3b
E/mali_gralloc(31877): ERROR: Unrecognized and/or unsupported format 0x3b and usage 0xb00
E/mali_gralloc(31877): ERROR: Format allocation info not found for format: 38
E/mali_gralloc(31877): ERROR: Format allocation info not found for format: 0
E/example.testapp(31877): Invalid base format! req_base_format = 0x0, req_format = 0x38
E/mali_gralloc(31877): ERROR: Unrecognized and/or unsupported format 0x38 and usage 0xb00
E/mali_gralloc(31877): ERROR: Format allocation info not found for format: 3b
E/mali_gralloc(31877): ERROR: Format allocation info not found for format: 0
E/example.testapp(31877): Invalid base format! req_base_format = 0x0, req_format = 0x3b
E/mali_gralloc(31877): ERROR: Unrecognized and/or unsupported format 0x3b and usage 0xb00
I/SurfaceView(31877): surfaceview vote min fps
D/BLASTBufferQueue(31877): [VRI[MainActivity]#2](f:0,a:1) acquireNextBufferLocked size=1084x2412 mFrameNumber=1 applyTransaction=true mTimestamp=397915572350486(auto) mPendingTransactions.size=0 graphicBufferId=136910672494629 transform=0
D/BLASTBufferQueue(31877): [SurfaceView[com.example.testapp/com.example.testapp.MainActivity]#3](f:0,a:1) acquireNextBufferLocked size=1084x2412 mFrameNumber=1 applyTransaction=true mTimestamp=397915580871178(auto) mPendingTransactions.size=0 graphicBufferId=136910672494621 transform=0
D/NtViewRootImpl(31877): mPopUpViewOffsets: offset=(0.0, 0.0), scale=(1.0, 1.0)
```

Besides the error above and the framerate issue, everything runs fine.

---
Flutter Version 3.29.0 stable
"Using the Impeller rendering backend (Vulkan)"
(profile mode)

**Comment by makoConstruct** [comment]
I'm using a Nothing CMF 2, confirm the above.

Wonderous for me is the only flutter app I've looked at that runs at 120fps, although it's extremely janky on the first taj mahal screen, which is suggestive that they've decided not to adopt impeller, and that's probably why their framerate is higher in other parts of the app.

**Comment by denizsafak** [comment]
@Sreeraj9764 I originally opened this as a separate issue #192600, and @puneetkukreja98 pointed out it's a duplicate of this one, so I'm posting my findings here too.

## What's going on?

Your Flutter app doesn't get stuck at 60Hz because of anything in your code. Android has a setting called `android:appCategory` that tells the phone what kind of app it is, and Flutter never sets it. Some phone makers use that setting, along with other signals, to guess whether an app is a game. Games push the GPU hard, which drains the battery fast and makes the phone heat up. To deal with that, **many manufacturers cap the refresh rate at 60Hz for anything they think is a game,** since a lower refresh rate means less GPU work and less heat during long sessions. **If the phone guesses wrong and flags your app as a game, it applies that same 60Hz cap to your app too, even though your app isn't doing anything game-like. And it applies that cap using a method that sits below anything your app can control.**

I dug into this for hours through display driver logs and decompiled system services to confirm it on Xiaomi/HyperOS. The same symptom has been reported on other brands too, including Nothing Phone, as @n7trd mentioned, so this likely isn't unique to Xiaomi, though I've only verified the exact mechanism on my own device.

## How it happens?

1. A Flutter app's manifest never sets `android:appCategory`. Left unset, Android's own default is `CATEGORY_UNDEFINED`, meaning the app has no declared category at all.
2. HyperOS (or your OS) scans newly installed/upgraded apps. If it sees an app rendering graphics with no declared category, it assumes the app might be a game and adds it to its own internal game list.
3. Once added, the phone's power management service sends a command directly to the display driver, telling it to cap the refresh rate at 60Hz.
4. I found this command by decompiling two system services: `com.xiaomi.joyose`, the **Game Turbo** service that manages game performance profiles, and `com.miui.powerkeeper`, the power management service (both keep these package names even on HyperOS). The call looked something like `setScreenEffect(pkg, 60, 254)`.
5. The priority value, 254, puts this command above WindowManager and above anything Flutter does at runtime. **That's why [flutter_displaymode](https://pub.dev/packages/flutter_displaymode) or [refresh_rate](https://pub.dev/packages/refresh_rate) does not work: your app is asking for 120Hz, but the phone already made its decision at a lower level and stopped listening.**

## The fix

Tell Android what kind of app you have. Add this to `AndroidManifest.xml`:

```xml
<application
    android:label="My App"
    android:name="${applicationName}"
    android:icon="@mipmap/ic_launcher"
    android:appCategory="productivity"> <!-- productivity, image, audio, video, social, news, maps -->
```

Once the phone knows the app isn't a game, the throttle clears.

>[!NOTE]
>**If your app was already added to Game Booster before this fix,** you'll also need to remove it manually: Game Booster → Settings/Add Games → turn it off. Other brands will have their own equivalent setting.

## Why Flutter shouldn't just hardcode the category?

Picking one category as the default would fix this for everyone, but **it would mislabel every Flutter app** that isn't a plain utility, games, social apps, media players, all of it, in the Play Store and the system app drawer. **This needs to be a choice each developer makes, not a template default.**

## What I'd suggest for Flutter?

1. Add an `--app-category` flag to `flutter create`, or prompt for it interactively, so it's a deliberate choice at project creation.
2. Add a commented-out `android:appCategory` line in `AndroidManifest.xml.tmpl` with a short note about the throttling risk, so developers see it during setup.

Both are manifest-level changes. Neither touches the engine or embedding layer, since the throttle happens entirely outside anything Flutter controls at runtime.

**Comment by makoConstruct** [comment]
Oh, I should also mention that a workaround was found, this works, at least on the Nothing cmf2.

https://pub.dev/packages/flutter_refresh_rate_control

```dart
await FlutterRefreshRateControl().requestHighRefreshRate();
```

This is a little difficult to square with the above interpretation, but anything is possible when it comes to third party battery life goodharting.

**Comment by n7trd** [comment]
@denizsafak 
Thanks, didn't know about `appCategory`! A flag or a hint, like you mentioned—would be really nice.

**Comment by sreeraj-gs** [comment]
@denizsafak Great work bro. The framework itself should set a default category or should ask for it
